services/api/kalvi/api/views/auth.py

```diff
@@ -1,15 +1,114 @@
+            serializer.is_valid(raise_exception=True)
+            email = serializer.validated_data['email']
+            password = serializer.validated_data['password']
+            user = authenticate(email=email, password=password)
+            if user is not None:
+                token = get_tokens_for_user(user)
+                return Response({'token' : token,'message' : 'Login Successful'}, status=status.HTTP_200_OK)
+            else:
+                #more specific error message to avoid revealing authentication mechanisms
+                error_message = "Invalid credentials, you need to register first."
+                return Response({'errors': error_message}, status=status.HTTP_401_UNAUTHORIZED)
+        except (ValidationError, PermissionError) as e:
+            return Response({'errors': str(e)}, status=status.HTTP_400_BAD_REQUEST)
 
+#viewset class for changing password, only authenticated users could change password. 
+class UserChangePasswordView(APIView):
+    renderer_classes = [UserRenderer]
+    permission_classes = [IsAuthenticated]
+    def post(self, request, format=None):
+        try:
+            serializer = UserChangePasswordSerialiser(data=request.data)
+            serializer.is_valid(raise_exception=True)
+            user = request.user
+            if not user:
+                return Response({'errors': "Unauthorised Access"}, status=status.HTTP_401_UNAUTHORIZED)
+            user.set_password(serializer.validated_data['password'])
+            user.save()
+            return Response({'msg': 'Password changed successfully'}, status=status.HTTP_200_OK)
+        except ValidationError as e:
+            return Response({'errors': e}, status=status.HTTP_400_BAD_REQUEST)
 
-class SignInEndPoint(APIView):
-    def post(self, request):
-        return Response("SIGNIN ENDPOINT")
+#viewset class for profile view of a user with get method using access tokens.
+class UserProfileView(APIView):
+    renderer_classes = [UserRenderer]
+    permission_classes = [IsAuthenticated]
+    def get(self, request, format=None):
+        try:
+            serializer = UserProfileSerialiser(request.user)
+            return Response(serializer.validated_data, status=status.HTTP_200_OK)
+        except ValidationError as e:
+            return Response({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+    
+#Viewset class for sending an email to reset the password in case of forgotten passwords.
+class SendPasswordResetEmailView(APIView):
+    renderer_classes = [UserRenderer]
+    def post(self, request, format=None):
+        try:
+            serializer = SendPasswordResetEmailSerializer(data=request.data)
+            serializer.is_valid(raise_exception=True)
+            return Response({'message': 'Password Reset link sent. Please check your Email'}, status=status.HTTP_200_OK)
+        except ValidationError as e:
+            return Response({'error': str(e)}, status=status.HTTP_400_BAD_REQUEST)
```

Comment:
should the status code 500 being used for generic errors rather than 400?

Reply:
exception here is catching serializer validation basically so incase it fails its bad request cz in sendpassword reset email serailizer all the request validation is happening , thats why i had kept bad request ,400......should i change it?

Reply:
let it be 400 bad request according to validation error in serializer